x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/Expression.java

```diff
@@ -99,13 +100,41 @@ public boolean childrenResolved() {
         return lazyChildrenResolved;
     }
 
+    /**
+     * Does the tree rooted at this expression have valid types at all nodes?
+     * <p>
+     *     For example, {@code SIN(1.2)} has a valid type and should return
+     *     {@link TypeResolution#TYPE_RESOLVED} to signal "this type is fine".
+     *     Another example, {@code SIN("cat")} has an invalid type in the
+     *     tree. The value passed to the {@code SIN} function is a string which
+     *     doesn't make any sense. So this method should return a "failure"
+     *     resolution which it can build by calling {@link TypeResolution#TypeResolution(String)}.
+     * </p>
+     * <p>
+     *     Take {@code SIN(1.2) + COS(ATAN("cat"))}, this tree should also
+     *     fail, specifically because {@code ATAN("cat")} is invalid. This should
+     *     fail even though {@code +} is perfectly valid when run on the results
+     *     of {@code SIN} and {@code COS}. And {@code COS} can operate on the results
+     *     of any valid call to {@code ATAN}. For this method to return a "valid"
+     *     result the <strong>whole</strong> tree rooted at this expression must
+     *     be valid.
+     * </p>
```

Comment:
Worth pointing out `childrenResolved` method to help with this process.
This extra check adds a lot of repetitive code through-out the code base:
```
if (childrenResolved() == false) {
   return new TypeResolution("children unresolved");
}
```
and we discussed into looking to remove it as it's defensive programming (and redundant) as the Verifiers already have this functionality in place.

Raised https://github.com/elastic/elasticsearch/issues/103834 to track it down.